File: leetcode_py/Clone Graph.py
# Definition for a undirected graph node
import new


class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

# depth first
class Solution:
    def cloneGraph(self, node):
        if node is None:
            return None

        start = UndirectedGraphNode(node.label)
        stack = [node]
        dicts = {node: start}
        while stack:
            current = stack.pop()
            for i in current.neighbors:
                if i not in dicts:
                    new = UndirectedGraphNode(i.label)
                    dicts[i] = new
                    stack.append(i)
                dicts[current].neighbors.append(dicts[i])
        return start

# breadth first
class Solution:
    def cloneGraph(self, node):
        if node is None:
            return None
        start = UndirectedGraphNode(node.label)
        dicts, current = {node: start}, [node]
        while current:
            next = []
            for i in current:
                for j in i.neighbors:
                    if j not in dicts:
                        new = UndirectedGraphNode(j.label)
                        next.append(j)
                        dicts[j] = new
                    dicts[i].neighbors.append(dicts[j])

            current = next
        return start


# recursive breadth first
class Solution:
    def deepcopy(self, node):
        if node.label in self.hash.keys():
            return self.hash[node.label]
        else:
            newnode = UndirectedGraphNode(node.label)
            self.hash[node.label] = newnode
            for i in node.neighbors:
                newnode.neighbors.append(self.deepcopy(i))
            return newnode

    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        self.hash = {}
        if node:
            return self.deepcopy(node)
        else:
            return None

File: leetcode_java/clone-graph/CloneGraph.java
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        UndirectedGraphNode result = null;
        if(node == null){
            return result;
        }
        Stack<UndirectedGraphNode> stack = new Stack<UndirectedGraphNode>();
        stack.push(node);
        List<Integer> list = new ArrayList<Integer>();
        HashMap<Integer,UndirectedGraphNode> map = new HashMap<Integer,UndirectedGraphNode>();
        UndirectedGraphNode front;
        while(!stack.empty()){
            front = stack.peek();
            stack.pop();
            if((front == null)||(list.contains(front.label))){
                continue;
            }
            list.add(front.label);
            if(!map.containsKey(front.label)){
                map.put(front.label,new UndirectedGraphNode(front.label));
            }
            if(result == null){
                result = new UndirectedGraphNode(front.label);
            }
            List<UndirectedGraphNode> neighbor = front.neighbors;
            for(UndirectedGraphNode nei:neighbor){
                stack.push(nei);
                if(!map.containsKey(nei.label)){
                    map.put(nei.label, new UndirectedGraphNode(nei.label));
                    
                }
                 map.get(front.label).neighbors.add(map.get(nei.label));
            }
        }
        return map.get(result.label);
    }
}